server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1871,26 +1877,67 @@ public void close() {
      */
     protected abstract void closeNoLock(String reason, CountDownLatch closedLatch);
 
+    protected final boolean isDrainedForClose() {
+        return ensureOpenRefs.hasReferences() == false;
+    }
+
+    protected final boolean isClosing() {
+        return isClosing.get();
+    }
+
+    protected final Releasable acquireEnsureOpenRef() {
+        if (isClosing() || ensureOpenRefs.tryIncRef() == false) {
+            ensureOpen(); // throws "engine is closed" exception if we're actually closed, otherwise ...
+            throw new AlreadyClosedException(shardId + " engine is closing", failedEngine.get());
+        }
+        return Releasables.assertOnce(releaseEnsureOpenRef);
+    }
+
+    protected final void drainForClose() {
+        if (isClosing.compareAndSet(false, true)) {
+            releaseEnsureOpenRef.close();
+        }
+        final var future = new PlainActionFuture<Void>() {
+            @Override
+            protected boolean blockingAllowed() {
+                // TODO remove this blocking, or at least do it elsewhere, see https://github.com/elastic/elasticsearch/issues/89821
+                return Thread.currentThread().getName().contains(ClusterApplierService.CLUSTER_UPDATE_THREAD_NAME)
+                    || super.blockingAllowed();
+            }
+        };
+        drainOnCloseListener.addListener(future);
+        try {
+            future.get();
+        } catch (ExecutionException e) {
+            logger.error("failure while draining operations on close", e);
+            assert false : e;
+            throw new IllegalStateException(e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            logger.error("interrupted while draining operations on close");
+            throw new IllegalStateException(e);
+        }
+    }
+
     /**
      * Flush the engine (committing segments to disk and truncating the
      * translog) and close it.
      */
     public void flushAndClose() throws IOException {
         if (isClosed.get() == false) {
-            logger.trace("flushAndClose now acquire writeLock");
-            try (ReleasableLock lock = writeLock.acquire()) {
```

Comment:
Besides draining, this write lock would also guarantee only one thread doing close or flushAndClose at the same time.

I wonder if we could let `drainForClose` return whether it was the first invocation (when it can set `isClosing` to true) and use this here (and in `close`) to ensure only the first invocation does anything? Or just use that to optimistically assert that there is only one thread closing this.